csharp/ql/src/semmle/code/csharp/Implements.qll

```diff
@@ -232,90 +233,158 @@ private Type getArgumentOrReturnType(Method m, int i) {
 }
 
 /**
- * Holds if `m2` is an interface method candidate for `m1`
- * (`m2 = getAnInterfaceMethodCandidate(m1)`), both `m1` and `m2` are
- * unbound generic methods, `t1` is a structural sub term of a
- * parameter type of `m1`, `t2` is a structural sub term of a parameter
- * (at the same index) type of `m2`, and `t1` and `t2` are compatible.
+ * INTERNAL: Do not use.
  *
- * Here, compatibility means that the two types are structurally equal
- * modulo identity conversions and method type parameters.
+ * Provides an implementation of Global Value Numbering for types
+ * (see https://en.wikipedia.org/wiki/Global_value_numbering), where
+ * types are considered equal modulo identity conversions and method
+ * type parameters (at the same index).
  */
-private predicate structurallyCompatible(
-  UnboundGenericMethod m1, UnboundGenericMethod m2, Type t1, Type t2
-) {
-  candidateTypes(m1, m2, t1, t2) and
-  (
-    // Base case: identity convertible
-    convIdentity(t1, t2)
-    or
-    // Base case: method type parameter (at same index)
-    exists(int i | structurallyCompatibleTypeParameter(m1, m2, i, t1, m2.getTypeParameter(i)))
-    or
-    // Recursive case
-    (
-      t1 instanceof PointerType and t2 instanceof PointerType
+module Gvn {
```

Comment:
This does seem like a complicated way to achieve this speedup, given that the net gains are fairly modest apart from fixing the timeout. Is there any way of addressing the timeout more directly?

Reply:
It does a bit more than just identity conversion, it also maps method type parameters at the same index to the same value. The reason that we need this is because we need to find compatible interface implementations:

```
interface I
{
    void M<T>(T[] x);
}

class A
{
    void M<T>(T[] x) { }
}

class B : A, I { }
```

The `overrides` relation generated by the extractor does not give us that `A.M()` may be a target in a call to `I.M()`, so we need to match the name and parameter types. If we do this with ordinary equality, `A.M()` will not match `I.M()`, because the two type parameters `T` in `I.M()` and `A.M()` are two different entities.